queue_listeners/lib/actionkit_page_parser.rb

```diff
@@ -0,0 +1,61 @@
+require_relative '../config/load_env'
+require_relative '../lib/crm_page'
+require 'json'
+
+class ActionkitPageParser
+  def initialize
+    @base_url = ENV['ACTIONKIT_BASE_URL']
+    @widget_to_page_types = {petition: 'Petition', donation: 'Donation'}
+  end
+
+  def parse_from_message(message)
+    parsed_json = JSON.parse(message)
+    page = CrmPage.new
+    page.base_url = @base_url
+
+    # We need to create CRM pages for each widget type that demands them, which might
+    # be a lot.
+    parsed_page = parsed_json['page']
+    widgets = parsed_page['widgets']
+    created_pages = []
+    if widgets
+      widgets.each do |widget|
```

Comment:
What if you filter out the relevant widgets:

```ruby
def relevant_widgets
  widgets.select{|widget| is_relevant?(widget)}
end

def is_relevant?(widget)
  !!WIDGET_TO_PAGE_TYPES[widget['widget_type'].to_sym]
end
```

You could then write `parse_from_message` as:

```ruby
def parse
  relevant_widgets.inject([]) do |pages, widget|
    pages << create_page_for_widget(widget)
    pages
  end
end
```